language/tools/move-package/src/resolution/resolving_table.rs

```diff
@@ -0,0 +1,187 @@
+    /// bindings between two account addresses that are unequal to each other), and succeeds
+    /// otherwise.
+    pub fn define(&mut self, name: QualifiedAddress, addr: Option<AccountAddress>) -> Result<()> {
+        let ix = self.get_or_create_assignment(name);
+        let Assignment::Assign(slot) = &mut self.assignments[ix] else {
+            unreachable!("Non-root assignment");
+        };
+
+        match (slot, addr) {
+            (_, None) => { /* nop */ }
+            (slot @ None, addr) => *slot = addr,
+            (Some(existing), Some(new)) => {
+                if *existing != new {
+                    return Err(unification_error(name.1, *existing, new));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Add the binding `a = b` to the table.  Fails if this introduces a contradiction (A path
+    /// through bindings between two account addresses that are unequal to each other), and succeeds
+    /// otherwise.
+    pub fn unify(&mut self, a: QualifiedAddress, b: QualifiedAddress) -> Result<()> {
+        let ix = self.get_or_create_assignment(a);
+        let jx = self.get_or_create_assignment(b);
+
+        let Some((ir, jr)) = self.get_mut_pair(ix, jx) else {
+            return Ok(())
+        };
+
+        let Assignment::Assign(ia) = ir else { unreachable!("Non-root assignment"); };
+        let Assignment::Assign(ja) = jr else { unreachable!("Non-root assignment"); };
+
+        match (ia, ja) {
+            (None, Some(_)) => *ir = Assignment::Linked(jx),
+
+            (Some(_), None) | (None, None) => *jr = Assignment::Linked(ix),
+
+            (Some(ia), Some(ja)) => {
+                if ia != ja {
+                    return Err(unification_error(a.1, *ia, *ja));
+                }
+            }
+        };
+
+        Ok(())
+    }
+
+    /// Returns the index of the "root" assignment (i.e. not a link to another assignment) for
+    /// `name` in this table, creating an empty assignment if one does not already exist.
+    ///
+    /// Performs path compression on the internal links to speed up future look-ups.
+    fn get_or_create_assignment(&mut self, name: QualifiedAddress) -> usize {
+        let Some(mut c) = self.redirection.get(&name).copied() else {
+            self.assignments.push(Assignment::Assign(None));
+            self.redirection.insert(name, self.assignments.len() - 1);
+            return self.assignments.len() - 1;
+        };
+
+        let Assignment::Linked(mut p) = self.assignments[c] else {
+            return c;
+        };
+
+        let Assignment::Linked(mut gp) = self.assignments[p] else {
+            return p;
+        };
+
+        let mut chain = vec![c];
+        while let Assignment::Linked(ggp) = self.assignments[gp] {
+            (c, p, gp) = (p, gp, ggp);
+            chain.push(c);
+        }
+
+        for link in chain {
+            self.assignments[link] = Assignment::Linked(gp);
+        }
```

Comment:
I'm not sure I understand why the updates are done here? Are they necessary or just an optimization?

Reply:
Just an optimisation, it's path compression just like for union find, it is a trick I learnt from one of Oleg Kiselyov's articles about implementing Algorithm W, a long time ago: https://okmij.org/ftp/ML/generalization.html (search for "path compression").